crates/zk-por-cli/src/verifier.rs

```diff
@@ -52,34 +53,63 @@ pub fn verify_user(
     let user_proof_paths =
         find_matching_files(user_proof_path_pattern).map_err(|e| PoRError::Io(e))?;
     let proof_file_num = user_proof_paths.len();
-    println!("successfully identify {} user proof files", proof_file_num);
+    if proof_file_num == 0 {
+        return Err(PoRError::InvalidParameter(format!(
+            "fail to find any user proof files with pattern {}",
+            user_proof_path_pattern
+        )));
+    }
+
+    if verbose {
+        println!("successfully identify {} user proof files", proof_file_num);
+    }
 
     let bar = ProgressBar::new(proof_file_num as u64);
     let chunk_size: usize = num_cpus::get();
-    user_proof_paths.chunks(chunk_size).for_each(|chunks| {
-        chunks.par_iter().for_each(|user_proof_path| {
-            let merkle_path = File::open(&user_proof_path).unwrap();
-            let reader = std::io::BufReader::new(merkle_path);
-            let proof: MerkleProof = from_reader(reader).unwrap();
-            if let Err(e) = proof.verify_merkle_proof(root_hash) {
-                panic!(
-                    "fail to verify the user proof on path {:?} due to error {:?}",
-                    user_proof_path, e
-                )
+    let invalid_proof_paths = user_proof_paths
+        .chunks(chunk_size)
+        .map(|chunks| {
+            let invalid_proof_paths = chunks
+                .par_iter()
+                .map(|user_proof_path| {
+                    let merkle_path = File::open(&user_proof_path).unwrap();
+                    let reader = std::io::BufReader::new(merkle_path);
+                    let proof: MerkleProof = from_reader(reader).unwrap();
+                    let result = proof.verify_merkle_proof(root_hash);
+                    (result, user_proof_path)
+                })
+                .filter(|(result, _)| result.is_err())
+                .map(|(_, invalid_proof_path)| invalid_proof_path.to_str().unwrap().to_owned())
+                .collect::<Vec<String>>();
+            if verbose {
+                bar.inc(chunks.len() as u64);
             }
```

Comment:
You dont need to manually select chunk size to be num cpus, the way the rayon threading works, it will automatically schedule num_cpus number of threads and will spin up a new thread every time one completes. Unless you specify the number of threads in a threadpool rayon will use num cpu threads at a time.

Reply:
cliff determines the chunk size as the number of cpus. So I think I will better keep it unchanged.

Reply:
Yeah but this version is more inefficient and perhaps cliff is unaware about this feature of rayon.

Reply:
https://developer.algorand.org/docs/get-details/transactions/#sending-a-transaction-in-the-future